Improve finite difference approximation for non-equidistant discretization step

vezeli opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Closely related to #17209
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Before the fix it was possible to assign undefined functions to ``points`` argument of ``as_finite_difference`` function but the undefined function was not discretized at the right point (which should be ``x0`` point). This fix corrects the behavior of undefined functions so that they are discretized mathematically in the right way.

Poor explanation but it is very simple. Here is hands-on example:
```
from sympy import *
f, dx = symbols('f dx', cls=Function)
x = Symbol('x')
expr = diff(f(x), x).as_finite_difference(points=dx(x), x0=x+1)
```
Solution before fix:
```
expr = -f(x - dx(x)/2 + 1)/dx(x) + f(x + dx(x)/2 + 1)/dx(x)
```
Solution after fix:
```
expr = -f(x - dx(x + 1)/2 + 1)/dx(x + 1) + f(x + dx(x + 1)/2 + 1)/dx(x + 1)
```


#### Other comments
Aside from being mathematically correct, this commit is the first step to solving #17209, as I see it. I have a working solution for that issue but I want to keep this commit as clean as possible since it is looks enough to be a stand-alone commit. Also I would like to test the second part of the solution a bit more.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Improved finite difference approximation for non-equidistant discretization steps.

<!-- END RELEASE NOTES -->

Commit:
Improve discretization for non-equidistant step

Adds a possibility of using undefined functions for finite difference
approximation.

oscarbenjamin commented:
Looks like there are some test failures. Maybe you need to import diff or use the `.diff` method instead.

Commit:
Correct dependency issue

vezeli commented:
Right, silly me... Sorry about that. Fixed it, I think it should work now.

vezeli commented:
Why does the coverage go down? I have introduced a test, shouldn't it go up?

oscarbenjamin commented on sympy/calculus/finite_diff.py:
```diff
@@ -387,6 +395,9 @@ def _as_finite_diff(derivative, points=1, x0=None, wrt=None):
     if not iterable(points):
+        if hasattr(points, 'is_Function'):
+            if points.is_Function and wrt in points.args:
+                points = points.subs(wrt, x0)
```
is_Function should be defined for all Basic instances. Why do we need to check if the attribute is present. Also you can do this more efficiently and cleanly with
```
if getattr(points, 'is_Function', False) and wrt in points.args:
```

vezeli replied:
>  Why do we need to check if the attribute is present?

I was worried that some users might set ``points`` to be Python integer (or float) and not Sympy, which would results in an unwanted error. By checking if it has ``is_Function`` the code does not break. You are right using `getattr` is a better idea, thanks.

What would be the best way to check if ``points`` is an object like `Function('dx')(Symbol('x')`?

smichr replied:
```python
>>> isinstance(f(x), Function) and len(f(x).args) == 1
True
```
That will also pass if `f` is `exp`.

sylee957 commented:
> Why does the coverage go down? I have introduced a test, shouldn't it go up?

It's because of the network issue, not from any changes made in here.

Commit:
Modify logics in if loop

sylee957 commented on sympy/calculus/finite_diff.py:
```diff
@@ -349,6 +349,14 @@ def _as_finite_diff(derivative, points=1, x0=None, wrt=None):
+    >>> dx = Function('dx')
+    >>> as_finite_diff(f(x).diff(x), points=dx(x), x0=x-h)
+    -f(-h + x - dx(-h + x)/2)/dx(-h + x) + f(-h + x + dx(-h + x)/2)/dx(-h + x)
+
```
I see this function is has public version in `Derivative.as_finite_diff`, and documentation updates may better go there.
I think the documentation here may completely be removed because this mostly contains duplicate information, with the only difference of using deprecated feature in examples.

vezeli replied:
OK, I see your point I will put documentation and example in functions.py.

Should I remove the documentation and keep the example in ``as_finite_diff`` function or remove both?

sylee957 replied:
Updates should go to `as_finite_difference` in `sympy/core/functions.py`.
I think that this doc can be removed because it is private function, and would set the docstring of the deprecated function `as_finite_diff.__doc__` as something like 
`"""Deprecated function. Use Diff.as_finite_difference instead"""` so that the sympy website no longer show another misguided doc in https://docs.sympy.org/latest/modules/calculus/index.html?highlight=as_finite_diff#sympy.calculus.finite_diff.as_finite_diff

vezeli replied:
Sounds good. I changed it.

Commit:
Change documentation

Since `_as_finite_diff` is a private function it is better to remove the
documentation and move it to `sympy/core/functions.py`. Additionally,
documentation for the deprecated function `as_finite_diff` is changed so
that the SymPy website no longer shows misguided doc.

vezeli commented:
Hey, what is the status of this PR? :)

smichr commented:
> Hey, what is the status of this PR? :)

It's in. :-)
